definitions/routing.lua

```diff
@@ -0,0 +1,249 @@
+-- move all of these down into rc config
+--
+-- or should these be put in routing lib?
+local TRACK_INFO_AUDIO_SRC_DISABLED = -1
+local TRACK_INFO_MIDIFLAGS_ALL_CH = 0
+local TRACK_INFO_MIDIFLAGS_DISABLED = 4177951
+local TRACK_INFO_CATEGORY_SEND = 0 -- send
+local TRACK_INFO_CATEGORY_RECIEVE = -1 -- send
+local TRACK_INFO_CATEGORY_HARDWARE = 1 -- send
+
+--//////////////////////////////////////////////////
+--
+-- ROUTING
+--
+--  neither of a and m | a and not m
+--      create AUDIO send (default)
+--
+--  only m
+--      create midi send
+--
+--  a and m
+--      create both audio and m
+--
+--  ----------------------------
+--
+--  the param_name is not necessarily the input format ?!
+--
+--  ----------------------------
+--  prepend key with `!` to update w/ disable value
+--
+--  `()u!a` = disablu audio for track ()
+--
+--  ----------------------------
+--  EXAMPLES
+--
+--  (176)               only audio send from sel to #176
+--
+--  (176)(50)a          only audio send from #176 to #50
+--
+--  (nameA)(nameB){2}   midi send from tr_nameA ch 0 to tr_nameV B ch 2
+--
+--  (nameB)[0|2]{}
+--        midi send from tr_nameA ch 0 to tr_nameV B ch 2
+--
+--
+--
+--
+--//////////////////////////////////////////////////
+
+return {
+
+  --  /////////////////////////////////////////////////////////////////////
+  --  ROUTING MACROS
+  --  ////////////////
+  --
+  --  TODO
+  --
+  --  - auto send (SRC_PATTERN) to (DEST_PATTERN)
+
+
+  macros = {
+    ['X'] = '(<dest_tr_name>)a2!m'
+  },
+
+  -- SET : always work with tables in the end
+  -- GET : for loop
+  --
+  src_guids = {},
+  dst_guids = {},
+  -- src_from_str = false,
+  -- src_from_sel = false,
+  dst_from_str = false,
+  user_input = false,
+
+  coded_targets = false,
+
+  -- code_tot_route_num_limit = 5,
+  tot_route_num_limit = 5,
+  route_num_extreme = 20,
+  remove_routes = false,
+
+
+  -- 0 send, 1 rec, 2 both (remove..)
+  -- you cannot create with `2`
+  category = 0,
+
+
+  --  /////////////////////////////////////////////////////////////////////
+  --  FLAGS
+  --  ///////
+
+
+  flags = {
+    AUDIO_SRC_OFF = TRACK_INFO_AUDIO_SRC_DISABLED,
+    MIDI_ALL_CH = TRACK_INFO_MIDIFLAGS_ALL_CH,
+    MIDI_OFF = TRACK_INFO_MIDIFLAGS_DISABLED,
+    CAT_SEND = TRACK_INFO_CATEGORY_SEND,
+    CAT_REC = TRACK_INFO_CATEGORY_RECIEVE,
+    CAT_HW = 1,
+  },
+
+  -- ////////////////////////////////////////////////////////////////////
+  -- PARAMS
+  -- ////////////
+  -- remove `param_` >>> becomes redundant
+
+  new_params = {--[[ empty ]]},
+
+  default_params = {
+    -- ////////////////////////////////////////////////////////////////////
+    -- WHICH TYPE
+    -- ////////////
+
+    ["a"] = {
+      description = 'SOURCE CHAN | int, index, &1024=mono, -1 for none',
```

Comment:
I'm not sure exactly what this does still, but I believe you have made a scheme for getting arbitrary user input in the middle of key sequences, cool!

Reply:
yeah I hope this can be a good foundation for doing all route stuff with keys. and i think that with your skills it
could turn out really great in the end.

Reply:
In rlib string is where i filter string for route params. I am excited to hear what u think of this. Maybe you can come up with an even better way of doing stuff but a lot of the hwavylifting i hope is done. I just pray to good the code is somewhat readable so that you can test it.

Reply:
Now I have been trying to make action that setsup automatic sidechain routes based on track names. Eg send all kick drums to ghost kick track and then sidechain all tracks that begin with pattern to ghost kick. This kind of auto route would allieviate a shitload of manual stuff. Allso i try to setup automatic submixes based on track names.

Reply:
You creates route in popup user input.
Think ex-commands in vim like :w :set xyz.

Reply:
I am not spending too much time on this project anymore so I won't be able to help implement any functionality. 

The routing system also feel a bit unnecessary to me, I think users would prefer to do routing manually because it is not a common operation and it is simpler for them that way. The definitions for routings are also complicated and may confuse users looking into the definitions folder in order to configure RK. I won't be merging this request for this reason. Feature bloat is also a real thing!

Reply:
I won't be merging this as it is right now for this reason. I may not merge it in the future after a lot of refactoring as well, so don't count on that! 

Thanks for your passion about Reaper Keys though, it's much appreciated to have someone else like this project as much as you do. If you would still like to help, I recommend trying to tackle some user issues like the confusion at the start and the completion window focus issue.

Reply:
Why are you not working on it so much anymore?
Yes I am going to look into those issues. I have to see how I can add them to my todo list.

Reply:
I'm developing a vision right now for a music interface project that is something like a Modular Digital Audio Workstation + a controller for it, which can optimally replace the function of the DAW and stay more true to music, real-time expression, and visions, and require less ego mental models to operate.

Reply:
I bet you would be interested seeing as how you love coding and music as much as I do.

Reply:
It's still in a watery intuition visionary stage in the pipeline of creation, so I've been holding off on sharing it for a while now. Here's a [demo](https://www.youtube.com/watch?v=FpHICKVpKT4) of an earlier version of the prototype for the controller I made which you may like.

Reply:
So I've been really excited about this project for a while now and haven't been spending a lot of time on Reaper Keys anymore.

Reply:
Dude I missed these comments lol. It's getting a bit cluttered.

Allright I see! I am going to checkout the demo. So could this program also be used together with reaper or is it like a full replacement sort of?

Reply:
Or would it make sense to reuse reaper keys to give you a modal keybingins system for switching between blocks etc?

Reply:
Would this be a more nische kind of app compared to reaper that woul maybe be more generalist kind of? I am not sure if I am at the level where I can even ask question about when and why to build a music app from scratch? I mean from my perspective it looks a little bit like a new gui that could be layered on top of reaper's audio engine or am I wrong?
Could I make drum beats with this as an example? I am probably asking some stupid questions since I don't know your full vision.

Reply:
could i use this instrument to record into reaper or would I do a full production with emergem blocks? What language do your write in?

Reply:
would this app be geared toward a particular genre? is it a midi instrument on it's own or do I have to work together with real instruments or midi instruments?

Reply:
What problems does this app solve that reaper doesn't?

Reply:
Hey I'm working on videos to communicate the concepts, I'll share with you when I'm done. There's a lot of content.

Reply:
holy shit. cool!